[SPARK-34638][SQL] Single field nested column prune on generator output

viirya opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://spark.apache.org/contributing.html
  2. Ensure you have added or run the appropriate tests for your PR: https://spark.apache.org/developer-tools.html
  3. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP][SPARK-XXXX] Your PR title ...'.
  4. Be sure to keep the PR description updated to reflect all changes.
  5. Please write your PR title to summarize what this PR proposes.
  6. If possible, provide a concise example to reproduce the issue for a faster review.
  7. If you want to add a new configuration, please read the guideline first for naming configurations in
     'core/src/main/scala/org/apache/spark/internal/config/ConfigEntry.scala'.
-->

### What changes were proposed in this pull request?
<!--
Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue. 
If possible, please consider writing useful notes for better and faster reviews in your PR. See the examples below.
  1. If you refactor some codes with changing classes, showing the class hierarchy will help reviewers.
  2. If you fix some SQL features, you can provide some references of other DBMSes.
  3. If there is design documentation, please add the link.
  4. If there is a discussion in the mailing list, please add the link.
-->

This patch proposes an improvement on nested column pruning if the pruning target is generator's output. Previously we disallow such case. This patch allows to prune on it if there is only one single nested column is accessed after `Generate`.

E.g., `df.select(explode($"items").as('item)).select($"item.itemId")`. As we only need `itemId` from `item`, we can prune other fields out and only keep `itemId`.

In this patch, we only address explode-like generators. We will address other generators in followups.

### Why are the changes needed?
<!--
Please clarify why the changes are needed. For instance,
  1. If you propose a new API, clarify the use case for a new API.
  2. If you fix a bug, you can clarify why it is a bug.
-->

This helps to extend the availability of nested column pruning.

### Does this PR introduce _any_ user-facing change?
<!--
Note that it means *any* user-facing change including all aspects such as the documentation fix.
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description and/or an example to show the behavior difference if possible.
If possible, please also clarify if this is a user-facing change compared to the released Spark versions or within the unreleased branches such as master.
If no, write 'No'.
-->

No

### How was this patch tested?
<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
-->

Unit test

viirya committed:
Nested column prune on generator output for one field.

SparkQA commented:
**[Test build #136541 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/136541/testReport)** for PR 31966 at commit [`5221be3`](https://github.com/apache/spark/commit/5221be35710ed149238cd801754790e7442f954b).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/41125/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/41125/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/41125/

SparkQA commented:
**[Test build #136541 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/136541/testReport)** for PR 31966 at commit [`5221be3`](https://github.com/apache/spark/commit/5221be35710ed149238cd801754790e7442f954b).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

viirya committed:
Add e2e test cases.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/136541/

SparkQA commented:
**[Test build #136549 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/136549/testReport)** for PR 31966 at commit [`4758f96`](https://github.com/apache/spark/commit/4758f96e788bf7c38f81bdf0603eb4493422764d).

viirya committed:
Fix wrong transform logic.

viirya commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala:
```diff
@@ -241,12 +262,69 @@ object GeneratorNestedColumnAliasing {
+          // For multiple field case, we cannot directly move field extractor into
+          // the generator expression. A workaround is to re-construct array of struct
+          // from multiple fields. But it will be more complicated and may not worth.
+          if (nestedFieldsOnGenerator.size == 1) {
```
For now I only add single field support. Multiple field case would be more complicated. We can consider to add it later.

maropu replied:
Could we file jira for this TODO task and leave the jira ID here?

viirya replied:
sure.

SparkQA commented:
**[Test build #136553 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/136553/testReport)** for PR 31966 at commit [`7908788`](https://github.com/apache/spark/commit/7908788447625229e4f6d71ab728f1a549857089).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/41134/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/41134/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/41134/

SparkQA commented:
**[Test build #136549 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/136549/testReport)** for PR 31966 at commit [`4758f96`](https://github.com/apache/spark/commit/4758f96e788bf7c38f81bdf0603eb4493422764d).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/136549/

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/41137/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/41137/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/41137/

SparkQA commented:
**[Test build #136553 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/136553/testReport)** for PR 31966 at commit [`7908788`](https://github.com/apache/spark/commit/7908788447625229e4f6d71ab728f1a549857089).
 * This patch **fails SparkR unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/136553/

viirya commented:
cc @cloud-fan @maropu @dongjoon-hyun

dongjoon-hyun commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasingSuite.scala:
```diff
@@ -336,7 +336,7 @@ class NestedColumnAliasingSuite extends SchemaPruningTest {
     val query = companies
       .generate(Explode('employers.getField("company")), outputNames = Seq("company"))
-      .select('company.getField("name"))
+      .select('company.getField("name"), 'company.getField("address"))
```
If we want to revise this test case, shall we revise the test name together?

viirya replied:
okay. revised.

dongjoon-hyun commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala:
```diff
@@ -241,12 +262,69 @@ object GeneratorNestedColumnAliasing {
+                // Replace the child expression of `ExplodeBase` generator with
+                // nested column accessor.
+                // E.g., df.select(explode($"items").as("item")) =>
+                //       df.select(explode($"items.a").as("item"))
```
The example description might look wrong because the column name is the same `item`, but the column type is not identical.
```scala
scala> sql("select explode(items.a) item from values array(named_struct('a', 1), named_struct('a', 2)) T(items)").show
+----+
|item|
+----+
|   1|
|   2|
+----+


scala> sql("select explode(items) item from values array(named_struct('a', 1), named_struct('a', 2)) T(items)").show
+----+
|item|
+----+
| {1}|
| {2}|
+----+
```

viirya replied:
oh, I miss the nested column accessor on top of it. So it looks not correct. I will update the comment.

dongjoon-hyun commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala:
```diff
@@ -241,12 +262,69 @@ object GeneratorNestedColumnAliasing {
+                    .map { case (oldAttr, newAttr) =>
+                  newAttr.withExprId(oldAttr.exprId).withName(oldAttr.name)
+                }
+                assert(updatedGeneratorOutput.length == rewrittenG.generatorOutput.length,
```
Just a question, according to the logic, is this the same with checking `rewrittenG.generator.elementSchema.toAttributes.length == rewrittenG.generatorOutput`?

viirya replied:
yea, i think this is the same.

dongjoon-hyun commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala:
```diff
@@ -241,12 +262,69 @@ object GeneratorNestedColumnAliasing {
+
+              case _ => pushedThrough
+            }
+          } else {
```
If you don't mind, shall we switch the body of `if` and `else`?

viirya replied:
okay

dongjoon-hyun reviewed (commented):
Thank you for pinging me, @viirya . I agree that we can handle `multiple field case` later.
> For now I only add single field support. Multiple field case would be more complicated. We can consider to add it later.

maropu commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala:
```diff
@@ -241,12 +262,69 @@ object GeneratorNestedColumnAliasing {
+                      case _: Attribute =>
+                        e.child
+                      case g: GetStructField =>
+                        ExtractValue(g.child, Literal(g.extractFieldName), SQLConf.get.resolver)
```
We need tests for case-sensitivity?

viirya replied:
let me add one.

viirya replied:
Not add this test case yet. Not related to this PR but seems there is a bug on case-sensitive nested column pruning. Let me figure it out first.

viirya replied:
Found a bug and proposed a fix at #32059. cc @maropu @dongjoon-hyun

viirya replied:
As #32059 was merged, I add case-sensitivity test.

maropu commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala:
```diff
@@ -231,6 +231,27 @@ object NestedColumnAliasing {
  */
 object GeneratorNestedColumnAliasing {
+  // Partitions `attrToAliases` based on whether the attribute is in Generator's output.
+  private def aliasesOnGeneratorOutput(
+      attrToAliases: Map[ExprId, Seq[Alias]],
+      generatorOutput: Seq[Attribute]) = {
+    val generatorOutputExprId = generatorOutput.map(_.exprId)
+    attrToAliases.partition { k =>
+      generatorOutputExprId.contains(k._1)
+    }
+  }
+
+  // Partitions `nestedFieldToAlias` based on whether the attribute of nested field extractor
+  // is in Generator's output.
+  private def nestedFieldOnGeneratorOutput(
+      nestedFieldToAlias: Map[ExtractValue, Alias],
+      generatorOutput: Seq[Attribute]) = {
+    val generatorOutputSet = AttributeSet(generatorOutput)
+    nestedFieldToAlias.partition { pair =>
+      pair._1.references.subsetOf(generatorOutputSet)
+    }
+  }
```
Since these functions are used only once, how about inlining them?

viirya replied:
okay for me. Put it as functions not for reuse but for making the code look simpler.

viirya replied:
If you don't mind, I prefer keep it as functions. This rule is already a bit complicated. Inlining these code make it looks harder to read.

maropu replied:
sgtm

maropu commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala:
```diff
@@ -241,12 +262,69 @@ object GeneratorNestedColumnAliasing {
-      NestedColumnAliasing.getAliasSubMap(exprsToPrune, g.qualifiedGeneratorOutput).map {
+      NestedColumnAliasing.getAliasSubMap(exprsToPrune).map {
         case (nestedFieldToAlias, attrToAliases) =>
           // Defer updating `Generate.unrequiredChildIndex` to next round of `ColumnPruning`.
```
Move this comment into L275-276?

maropu reviewed (commented):
This behaivour (the col name that is internally auto-generated is visible to a user) is expected one?
```
scala>  sql("select __auto_generated_subquery_name.col.a from (select explode(items) from t)").show()
+---+
|  a|
+---+
|  1|
|  2|
+---+
```

viirya commented:
> This behaivour (the col name that is internally auto-generated is visible to a user) is expected one?
> 
> ```
> scala>  sql("select __auto_generated_subquery_name.col.a from (select explode(items) from t)").show()
> +---+
> |  a|
> +---+
> |  1|
> |  2|
> +---+
> ```

I can see we use it in test:
https://github.com/apache/spark/blob/0494dc90af48ce7da0625485a4dc6917a244d580/sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala#L2788

viirya committed:
Address comments.

viirya commented:
Addressed above comments except for https://github.com/apache/spark/pull/31966#discussion_r606961206. For case-sensitive test case, I need to figure out a suspicious bug in nested column pruning first.

SparkQA commented:
**[Test build #136907 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/136907/testReport)** for PR 31966 at commit [`fe286df`](https://github.com/apache/spark/commit/fe286df4a856bd1fcf162794c543e84d965dce2e).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/41484/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/41484/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/41484/

SparkQA commented:
**[Test build #136907 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/136907/testReport)** for PR 31966 at commit [`fe286df`](https://github.com/apache/spark/commit/fe286df4a856bd1fcf162794c543e84d965dce2e).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/136907/

viirya committed:
Merge remote-tracking branch 'upstream/master' into SPARK-34638

viirya committed:
Add case-insensitive test.

viirya committed:
Add comment.

viirya commented:
@maropu @dongjoon-hyun Thanks for your review and patience. I added test-sensitivity test. Once you have time, please take another look. Thanks.

SparkQA commented:
**[Test build #137170 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/137170/testReport)** for PR 31966 at commit [`6c9d839`](https://github.com/apache/spark/commit/6c9d839a53095cdd24abe2007a1576a7c80a8f93).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/41748/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/41748/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/41748/

SparkQA commented:
**[Test build #137170 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/137170/testReport)** for PR 31966 at commit [`6c9d839`](https://github.com/apache/spark/commit/6c9d839a53095cdd24abe2007a1576a7c80a8f93).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/137170/

maropu commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala:
```diff
@@ -241,12 +262,68 @@ object GeneratorNestedColumnAliasing {
+                // E.g., df.select(explode($"items").as("item")).select($"item.a") =>
+                //       df.select(explode($"items.a").as("item.a"))
+                val rewrittenG = newG.transformExpressions {
+                  case e: ExplodeBase =>
```
This PR targets at explode-like generators only? How about the others, e.g., `inline`?

viirya replied:
Ideally it is okay too. We also support other generator pruning through (plz see `canPruneGenerator`). As the reported case is `explode`, I work on it first. We can add more support in follow-ups.

maropu replied:
yea, targeting `explode` in this PR sounds okay. Could you write it explicitly in the PR description?

viirya replied:
Thanks. Added a note in the description.

maropu commented on sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/SchemaPruningSuite.scala:
```diff
@@ -351,6 +351,41 @@ abstract class SchemaPruningSuite
 
+  testSchemaPruning("SPARK-34638: nested column prune on generator output") {
+    val query1 = spark.table("contacts")
+      .select(explode(col("friends")).as("friend"))
```
Can this work correctly if `explode` has another func in its child, e.g., `explode(filter(friends, x -> x is not null))`?

viirya replied:
Let me add a test to verify it.

viirya replied:
Oh, because in the lambda function you access `friends`, Spark cannot prune it.

viirya replied:
The higher order function expression is a special case. I just looked at it with more details. For example, `ArrayFilter` holds the complex expression `friends`, so from the point of view of Spark, `friends` is thought to be used by `ArrayFilter`.

I think it has some space to optimize it more, targeting higher order function expression. It is out of the scope of this PR, however. I will take some time to look at this after.

viirya commented on sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/SchemaPruningSuite.scala:
```diff
@@ -351,6 +351,41 @@ abstract class SchemaPruningSuite
+    checkScan(query3, "struct<friends:array<struct<first:string,middle:string,last:string>>>")
+    checkAnswer(query3, Row("Susan", "Z.", Row("Susan", "Z.", "Smith")) :: Nil)
+
+    withSQLConf(SQLConf.CASE_SENSITIVE.key -> "false") {
+      val query4 = spark.table("contacts")
+        .select(explode(col("friends")).as("friend"))
+        .select("friend.First")
+      checkScan(query4, "struct<friends:array<struct<first:string>>>")
+      checkAnswer(query4, Row("Susan") :: Nil)
+
+      val query5 = spark.table("contacts")
+        .select(explode(col("friends")).as("friend"))
+        .select("friend.MIDDLE")
+      checkScan(query5, "struct<friends:array<struct<middle:string>>>")
+      checkAnswer(query5, Row("Z.") :: Nil)
+    }
+  }
```
Sure.

viirya commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala:
```diff
@@ -241,12 +262,68 @@ object GeneratorNestedColumnAliasing {
+                      .getOrElse(f)
+                }
+
+              case _ => pushedThrough
```
okay.

viirya commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala:
```diff
@@ -241,12 +262,68 @@ object GeneratorNestedColumnAliasing {
+          // the generator expression. A workaround is to re-construct array of struct
+          // from multiple fields. But it will be more complicated and may not worth.
+          // TODO(SPARK-34956): support multiple fields.
+          if (nestedFieldsOnGenerator.size > 1 || nestedFieldsOnGenerator.size == 0) {
```
done.

viirya committed:
Address review comments.

SparkQA commented:
**[Test build #137407 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/137407/testReport)** for PR 31966 at commit [`ad3d191`](https://github.com/apache/spark/commit/ad3d19145e984a001205920ffd8dfde1f83ea4d1).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/41982/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/41982/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/41982/

SparkQA commented:
**[Test build #137407 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/137407/testReport)** for PR 31966 at commit [`ad3d191`](https://github.com/apache/spark/commit/ad3d19145e984a001205920ffd8dfde1f83ea4d1).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/137407/

maropu commented:
Looks fine otherwise. I think it's better to have @dongjoon-hyun check this before merging.

viirya commented:
Thanks @maropu!

dongjoon-hyun commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala:
```diff
@@ -241,12 +262,70 @@ object GeneratorNestedColumnAliasing {
+
+                // As we change the child of the generator, its output data type must be updated.
+                val updatedGeneratorOutput = rewrittenG.generatorOutput
+                    .zip(rewrittenG.generator.elementSchema.toAttributes)
```
Indentation? It seems that two-space is enough in this case.

dongjoon-hyun commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala:
```diff
@@ -241,12 +262,70 @@ object GeneratorNestedColumnAliasing {
+                  newAttr.withExprId(oldAttr.exprId).withName(oldAttr.name)
+                }
+                assert(updatedGeneratorOutput.length == rewrittenG.generatorOutput.length,
+                  "Updated generator output must have same length as original generator output.")
```
Maybe, `same length as` -> `the same length with`?

dongjoon-hyun commented:
Hi, @viirya . While testing this PR, I found the following regression.

**BEFORE (3.1.1)**
```scala
scala> sql("select * from values array(array(named_struct('a', 1, 'b', 3), named_struct('a', 2, 'b', 4))) T(items)").write.parquet("/tmp/nested_array")

scala> spark.read.parquet("/tmp/nested_array").createOrReplaceTempView("t")

scala> sql("select d.a from (select explode(c) d from (select explode(items) c from t))").show()
+---+
|  a|
+---+
|  1|
|  2|
+---+
```

**BEFORE(master)**
```scala
scala> spark.read.parquet("/tmp/nested_array").createOrReplaceTempView("t")

scala> sql("select d.a from (select explode(c) d from (select explode(items) c from t))").show()
+---+
|  a|
+---+
|  1|
|  2|
+---+
```

**AFTER (This PR)**
```scala
scala> spark.read.parquet("/tmp/nested_array").createOrReplaceTempView("t")

scala> sql("select d.a from (select explode(c) d from (select explode(items) c from t))").show()
java.lang.ClassCastException: org.apache.spark.sql.types.ArrayType cannot be cast to org.apache.spark.sql.types.StructType
  at org.apache.spark.sql.catalyst.expressions.SelectedField$.selectField(SelectedField.scala:81)
  at org.apache.spark.sql.catalyst.expressions.SelectedField$.unapply(SelectedField.scala:62)
  at org.apache.spark.sql.catalyst.expressions.SchemaPruning$.getRootFields(SchemaPruning.scala:124)
  at org.apache.spark.sql.catalyst.expressions.SchemaPruning$.$anonfun$identifyRootFields$1(SchemaPruning.scala:81)
```

Could you double-check this and add some test coverage?

viirya commented:
Thanks @dongjoon-hyun. Let me take a look.

viirya committed:
Deal with special cases.

viirya commented:
@dongjoon-hyun Sorry for late. I addressed the cases and added test for it.

viirya committed:
Address some comments.

SparkQA commented:
**[Test build #137898 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/137898/testReport)** for PR 31966 at commit [`a719409`](https://github.com/apache/spark/commit/a7194092ebc3a9ca97daba29fda0fcc0ec099cf4).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/42423/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/42423/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/42423/

dongjoon-hyun commented:
No problem. Thank you for updates, @viirya .

dongjoon-hyun commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala:
```diff
@@ -276,6 +276,16 @@ object GeneratorNestedColumnAliasing {
+          // It is because we don't allow field extractor on two-level array,
+          // i.e., attr.field when attr is a ArrayType(ArrayType(...)).
+          // Similarily, we also cannot push through if the child of generator is `MapType`.
+          g.generator.children.head.dataType match {
```
Got it. Let me play more with this PR for a while. It seems I need more tests.

viirya replied:
Thank you!

SparkQA commented:
**[Test build #137898 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/137898/testReport)** for PR 31966 at commit [`a719409`](https://github.com/apache/spark/commit/a7194092ebc3a9ca97daba29fda0fcc0ec099cf4).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/137898/

dongjoon-hyun reviewed (approved):
+1, LGTM. Thank you, @viirya and @maropu .

viirya commented:
Thank you @dongjoon-hyun @maropu

viirya commented:
Merging to master.
